### ISSUE TITLE
APPCRASH on small 20 page document duing final step

### ISSUE THREAD

**Issue by stguitar** [wkhtmltopdf/wkhtmltopdf#2894]
Ive been looking around today and have not been able to find anything about this issue I am seeing. 

I have a document that is about 10 webpages that gets glued together into one big HTML page.  The resulting PDF should be about 22 or 23 pages, and the filesize of the source concatenated HTML file is just under 1MB.

I have this process building other files with even more pages just fine, but for some reason this one is crashing WKHTMLTOPDF (windows).

It only seems to crash when trying to use a local HTML file for the footer.  If I dont use --footer-html then I do not see the issue.

I also do not see the issue at all on Mac, only my windows server.  I am running 12.3.2 with patched qt.

![screen shot 2016-04-18 at 2 12 53 pm](https://cloud.githubusercontent.com/assets/607525/14615510/f7da0518-0573-11e6-9719-92ec7b7333e6.png)
![screen shot 2016-04-18 at 2 13 57 pm](https://cloud.githubusercontent.com/assets/607525/14615509/f7d9b02c-0573-11e6-92e1-6fe3a77f17f1.png)

the final page total count seems off in the screenshot as well, perhaps this is related?

A little bit about the source HTML, it does reference JS/CSS/Image resources via a public domain name, but the same machine is hosting those resources.  I have tried to use local file paths for that even and still didnt work.

**Comment by stguitar** [comment]
playing around some more, when i remove --disable-smart-shrinking, the issue goes away.  is there a reason this is happening?

**Comment by ashkulz** [comment]
Can you post a sample test case which can reproduce the issue?

**Comment by stguitar** [comment]
Hey there, thanks for the reply. 

So I am trying to see if I can get you such a test.  The problem is that the source HTML contains client data that I cannot hand over.  It is also a rather large HTML document to comb through.  I have tried to run this same HTML through an older version of WK (0.11.0 rc2) and it works fine.  I am only seeing the issue on  12.3.2 + (saw it on the current alpha for 13 as well)

We are in the process of upgrading our WK with a new  version of our software that we are beta testing now to incorporate alot of the other positive changes to WK with the more recent versions.  

I havent seen this appcrash in any of our tests except for this one case, which is why i have a harder time in getting a good dummy test for you.

**Comment by stguitar** [comment]
@ashkulz I was able to get a generic enough HTML file for this failure, and am including it in a zip here.  The zip contains everything it needs to make the page viewable/exportable (referenced css files, referenced footer sample).  I am not sure if all that is needed but wanted to give it just in case. 

The command options that I am using are:

    -L 10mm -R 10mm -T 12mm -B 12mm --page-size Letter --enable-internal-links --disable-smart-shrinking --orientation Landscape --zoom 0.8 --footer-spacing 15 --footer-html tmp_footer.html --user-style-sheet print.css output_mod.html thedoc.pdf

Again, I cannot reproduce this on an older version, or on mac.  I also cannot reproduce if i remove the 'disable smart shrinking' flag, or the 'user-style-sheet' command, but I actually need to use both of those. Our server that will be generating these documents is a windows machine, so that is the other reason why I am seeing this error. 

Any help would be great!

[Archive.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/237167/Archive.zip)

**Comment by stguitar** [comment]
Oh, and when you run it, you will notice some warnings like:

    Warning: Received createRequest signal on a disposed ResourceObject's NetworkAccessManager. This might be an indication
of an iframe taking too long to load.

but those seem to happen even when I am able to generate properly using these sample/test documents - so those errors are probably safe to ignore.

**Comment by ashkulz** [comment]
It crashes for me on Windows as well. I guess you forgot to include `tmp_footer.html` in the archive, can you post that as well?

**Comment by stguitar** [comment]
@ashkulz sorry about that.  i thought i had included that already (well I meant to anyway) but in any case here it is. 

[tmp_footer.html.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/238663/tmp_footer.html.zip)

**Comment by ashkulz** [comment]
Hmm, this change seems to fix it. Can you compile and see if it works and generates things in your environment correctly?
```diff
diff --git a/src/lib/pdfconverter.cc b/src/lib/pdfconverter.cc
index 7a14e25..0c57627 100644
--- a/src/lib/pdfconverter.cc
+++ b/src/lib/pdfconverter.cc
@@ -995,12 +995,7 @@ void PdfConverterPrivate::printDocument() {
 		pageNumber=1;
 		for (int d=0; d < objects.size(); ++d) {
 			beginPrintObject(objects[d]);
-			// XXX: In some cases nothing gets loaded at all,
-			//      so we would get no webPrinter instance.
-			int pageCount = webPrinter != 0 ? webPrinter->pageCount() : 0;
-			//const settings::PdfObject & ps = objects[d].settings;
-
-			for(int i=0; i < pageCount; ++i) {
+			for(int i=0; i < objects[d].pageCount; ++i) {
 				if (!objects[d].headers.empty())
 					handleHeader(objects[d].headers[i], i);
 				if (!objects[d].footers.empty())
```

**Comment by stguitar** [comment]
That is great news.  However I have to admit I am no C programmer (anymore). 

I am happy to try and compile from source, are there any guides that I could follow to do that?

**Comment by ashkulz** [comment]
see the [compilation instructions](https://github.com/wkhtmltopdf/wkhtmltopdf/blob/master/INSTALL.md#windows) -- let me know if you face an issue.

**Comment by stguitar** [comment]
@ashkulz Fantastic.  I will let you know!

**Comment by stguitar** [comment]
@ashkulz just so you know, i am in progress with this.  it will take me a bit to get a windows development environment spun up but im working through it. im excited to help!

**Comment by stguitar** [comment]
@ashkulz I can concur that this seems to allow my document to generate. I am going to install this build on my actual server and give it a go with the real document next. 

if all is well there, what would be the next steps?

**Comment by ashkulz** [comment]
I can commit this patch, but I'm not sure that the fix is correct. It seems that there seems to be a difference in the number of pages counted at different times, and using one consistently fixes the crash.

However, does that cause the output to be truncated or having an extra blank page at the end? I wasn't able to get any text in your sample due to missing fonts, hence the request to verify whether the output is correct.

**Comment by stguitar** [comment]
gotcha.  I figured - my quick test yesterday showed me the same thing.  I am planning to run this on my actual document today and I will let you know. 

Thansk for the help and quick turn around.

**Comment by stguitar** [comment]
@ashkulz OK, so some strange results here.

first, let me say that we produce a lot of various PDF documents using WKHTMLTOPDF without error.  When we saw this problem, it was strange, and was only happening in this one document.  It happened every time, but it was confined to this one document.  We had the client disable the footers to get around it in the meantime.

For whatever reason, as of now, I am not able to get the same source document to fail. My sample I sent you still fails, but the real source that I created the sample off of isnt failing anymore. 

This patch doesnt make it work differently, I just cannot reproduce the source issue anymore. 

So, im not really sure what the next steps should be now.  I can saw that from what I could tell in the sample, i didnt see anything wrong despite the missing fonts.

**Comment by stguitar** [comment]
come to think on it, the samples i provided that you were working with, did you get a footer on those? i dont think i did even with the patch.

**Comment by ashkulz** [comment]
If the behavior is not different before/after, I'm more inclined to commit this patch (seems to have a lot of duplicates). Will wait for confirmation from at least one other duplicate issue.

**Comment by stguitar** [comment]
Cool, ill keep my eyes on its update.  I will say that we had another client in the past few days that had something similar happen, but I havent had a chance yet to dig in enough to determine if its the same thing or not.  I suspect it is, based on the description Ive heard so far. When I do, I will use the patched version of this to test it and see what happens.  I will let you know.

Thanks for all the help!

**Comment by mikecmpbll** [comment]
compiling it for trusty right now.. will let you know.

**Comment by taezaz** [comment]
Hi,

I had the same type of crash. After compiling with your patch wkhtmltopdf no longer segfaults, but there's a blank page at the end. It might be caused by something specific in my report, but I doubt it, since there was no blank page with old wkhtmltopdf in the variations of the report that worked. Everything else seems fine.

**Comment by ashkulz** [comment]
@taezaz: that's exactly what I was looking for! It seems that it is counting the number of pages differently, and in your case there's a blank page at the end ([as mentioned above](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2894#issuecomment-215628708)). Can you post a testcase where this happens, along with command-line so that I can verify what the ideal fix should be?

**Comment by stguitar** [comment]
just to revisit this, i had another document fail.  This document was actually fairly big (23 GB HTML document).  It failed in the same way. I did go through and scrub the file for content so I could share it here, but after spending time doing that, as a sanity check, i ran the originally failing document back through WK and it worked. So, I am no longer able to replicate with that file on the current live release.  Once I can get it to fail again, I plan to run it through my patched version with this change.

**Comment by ashkulz** [comment]
Do you really have a 23GB document or is it just 23MB? I doubt the former will work unless you have at least 128GB of RAM :smile: To me, there seems to be an issue in the counting pages and need to have at least one case why there seems to be a miscount to identify the correct fix.

**Comment by ashkulz** [comment]
While the above patch fixes the crash, that doesn't mean a blank page at end or missing page end can be ruled out -- which seems to be possible as per the comment from @taezaz

**Comment by stguitar** [comment]
You are correct! Sorry about that - too much going on over here.  23 MB file...

**Comment by rahult** [comment]
How do I get this patch? I cannot find it anywhere!

**Comment by gmillsap** [comment]
This issue happens to us on very rare occasions (probably less than .01%). However when it does happen it completely corrupts whatever printing process is being run. 

We are toying around with retrying the print, including a custom style sheet that alters the the css in a minor way to change how wkhtml is calculating the pages. It's not a perfect solution though, and it would be great if this bug was fixed.

**Comment by ashkulz** [comment]
Need a reproducible test case to come up with a proper fix!

**Comment by stguitar** [comment]
We have been doing a lot of updates to our application and have just been running of the current release for windows and I havent really seen this in a long while.  I am not sure I can come up with the exact case that I was originally seeing, but I will see what I can do.  If anyone else has a test case to share that may be quicker. 

When I get some time, I will try to re-step through what I originally posted and see if I can still get it to happen.

**Comment by gmillsap** [comment]
I wish I could help. This has been happening a lot to us lately. Well...we get 4-6 cases a month over a few hundred thousand documents. So maybe a lot is a stretch. 

But this is with 6 separate CSS hack files included that change the page dimensions ever so slightly in an attempt to alter the contents enough (yet subtly enough) to prevent the seg fault. Without our iterative process of attempting to fix the css...seg fault...attempt to fix...etc. We see about 100 of these over 200k documents.

All of our particular cases contain sensitive data that we cannot share, and simply changing the data on the document fixes the issue.

**Comment by ashkulz** [comment]
Well, the above patch will fix the issue but can result in an extra page. Unless I can figure out why the number of pages is changing, I can't really make a proper fix -- and I need **some** test case..

**Comment by gmillsap** [comment]
@ashkulz 

I finally came up with one of the issues and have it all pulled out into html. The command being used to print is:

wkhtmltopdf --disable-forms --page-size Letter --margin-bottom 17.7mm --margin-top 18mm --header-spacing 19 --footer-spacing 0 --encoding utf-8 --title "Bad Print" --header-html "header.html" --footer-html "footer.html" bad_print.html bad_print.pdf
[bad_print.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/651984/bad_print.zip)

**Comment by ashkulz** [comment]
Which platform ie OS, version, wkhtmltopdf version?

**Comment by gmillsap** [comment]
Happens on CentOS 6/7 and OSX, version 0.12.2.1 with patched qt

**Comment by gmillsap** [comment]
Also tried with 0.12.3 with patched qt on OSX and it failed as well.

**Comment by omh** [comment]
I also experienced issues with the occasional segfault with version 0.12.4. 2 out of ~300 PDFs would segfault as the page count was incorrect.

> Segmentation fault (core dumped)=============================] Page 5 of 4
 
I could work around it by disabling any margins and also by not using `--disable-smart-shrinking`, but that would mess up the layout.

Just tried the patch for a few PDFs that were segfaulting for me and it worked great for me.

System: Ubuntu 16.04.2

**Comment by hugo-ruivo-dw** [comment]
I am getting the same problem randomly crashes with the same exception code when generating in command line when executing with:

wkhtmltopdf.exe --header-html header.html --header-spacing 10 --dpi 96 --disable-smart-shrinking --enable-local-file-access --load-error-handling ignore --no-stop-slow-scripts --orientation Landscape --page-width 297 -B 10 -L 0 -R 0 -T 40 --zoom 1 body.html out.pdf

I am using version 0.12.3.1

I have checked out and built from source code in Windows and the crash occurs too, but it is not as often as with version 0.12.3.1, for example in 10 runs with version 0.12.3.1 crashes 9 times but source build crashes 1 times only.

Seems related to page count and specific source HTML code.

I ended up to apply something similar as in pseudocode from: #3366

**Comment by ashkulz** [comment]
So, in #2711 someone gave a reproducible test case under docker and I was able to figure out **why**  the page count was changing and fixed it in 6f77c4640a760d1297ad4d3441c12b8e9651245a. I've triggered a `-rc` build which should include this fix, please follow the other ticket for updates.
